frame/benchmarking/README.md

```diff
@@ -1,3 +1,182 @@
+This benchmarking framework will result in a file that automatically generates those formulas for
+you which you can then use in your pallet.
+
+## Writing Benchmarks
+
+Writing a runtime benchmark is much like writing a unit test for your pallet. They need to be
+carefully crafted to execute a certain logical path in your code. In tests you want to check for
+various success and failure conditions, but with benchmarks you specifically want to test for the
+**most computationally heavy** path, a.k.a the "worst case scenario".
+
+This means that if there are certain storage items or runtime state that may effect the complexity
+of the function, for example triggering more iterations in a `for` loop, to get an accurate result,
+you must set up your benchmark to trigger this.
+
+It may be that there are multiple paths your function can go down, and it is not clear which one is
+most heavy. In this case, you should just create a benchmark for each scenario! You may find that
+there are paths in your code where complexity may become unbounded depending on user input. This may
+hint to you that you should implement sane boundaries for how a user can use your pallet. For
+example: limiting the number of elements in a vector, limiting the number of iterations in a `for`
+loop, etc...
+
+Examples of end to end benchmarks can be found in the [pallets provided by Substrate](../), and the
+specific details on how to use the `benchmarks!` macro can be found in [its
+documentation](./src/lib.rs).
+
+## Testing Benchmarks
+
+You can test your benchmarks using the same test runtime that you created for your pallet's unit
+tests. By creating your benchmarks in the `benchmarks!` macro, it automatically generates test
+functions for you:
+
+```rust
+fn test_benchmark_[benchmark_name]<T>::() -> Result<(), &'static str>
+```
+
+Simply add these functions to a unit test and ensure that the result of the function is `Ok(())`.
+
+> **Note:** If your test runtime and production runtime have different configurations, you may get
+different results when testing your benchmark and actually running it.
+
+In general, benchmarks returning `Ok(())` is all you need to check for since it signals the executed
+extrinsic has completed successfully. However, you can optionally include a `verify` block with your
+benchmark which can additionally verify any final conditions, such as the final state of your
+runtime.
+
+These additional `verify` blocks will not effect the results of your final benchmarking process.
+
+To run the tests, you need to be sure to enable the `runtime-benchmarks` feature flag. This may also
+mean you need to move into your node's binary folder. For example, with the Substrate repository,
+this is how you would test the Balances Pallet's benchmarks:
+
+```bash
+cd bin/node/cli
+cargo test -p pallet-balances --features runtime-benchmarks
+```
+
+## Adding Benchmarks
```

Comment:
I think a logical ordering is first `## Adding Benchmarks` and then `## Testing Benchmarks`.